### PR TITLE
Updating Ip fields to use doc_values to search

### LINKED ISSUE #11144: `IpFieldMapper` should use `IndexOrDocValuesQuery` where possible
**Is your feature request related to a problem? Please describe.**
While going through the `IpFieldMapper` class and looking through `IpFieldMapperType` I noticed that the termQuery, termsQuery, and rangeQuery all use underlying lucene index queries. The documentation also incorrectly points to the fact that IP address do not have doc_values enabled by default. This is incorrect, ref: https://github.com/opensearch-project/OpenSearch/blob/b78176afefd8c09a19de6bc52907b40c8c4fedf2/server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java#L89

**Describe the solution you'd like**
Wrap all three queries in `IndexOrDocValuesQuery` and add checks to ensure that we return the index query if the field is only indexed and doc_values query if we only have doc_values enabled.

**Describe alternatives you've considered**
Alternative is to leave this as is. But there is potentially a lot of performance improvement to be had. Consider a query that looks like

```
{
  query: {
  "range":{
  "timestamp: {
      "from" :xyz,
       "to": xyz
    }
   },
"terms": {
"ipaddress" :["0.0.0.0", "1.1.1.1"]
}
}
}
```
We could let the range on the timestamp drive the query and use the doc_values to match the documents for the terms in the ipaddress.

**Additional context**
#7057

### PR DESCRIPTION
**Pull request by harshavamsi**
### Description
Fixes `term` and `range` queries on `IpFields` to be able to perform searches when only `doc_values` is enabled.

### Related Issues
Resolves #11144 

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [X] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by harshavamsi** [comment]
https://github.com/opensearch-project/OpenSearch/issues/11547 -- flaky test

File: server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java [hunk 1]
```diff
@@ -222,25 +224,51 @@ protected Object parseSourceValue(Object value) {
...
+                    InetAddress address = InetAddresses.forString(term);
+                    query = InetAddressPoint.newExactQuery(name(), address);
                 }
-                InetAddress address = InetAddresses.forString(term);
-                return InetAddressPoint.newExactQuery(name(), address);
             }
+            if (isSearchable() && hasDocValues()) {
+                return new IndexOrDocValuesQuery(
+                    query,
+                    SortedSetDocValuesField.newSlowRangeQuery(
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java:247]: @harshavamsi this is a single term query, shouldn't we use `SortedSetDocValuesField.newSlowExactQuery` instead?

Comment (harshavamsi) [reply server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java:247]: ah you're right

**Review by reta — APPROVED** [review]

**Comment by harshavamsi** [comment]
@reta fixed conflicts and some failing tests. @msfroh could you also take a look

**Comment by msfroh** [comment]
Getting a failure from forbiddenApis (which we should be able to catch with `./gradlew precommit`):

```
> Task :server:forbiddenApisMain
Forbidden method invocation: java.lang.String#getBytes() [Uses default charset]
  in org.opensearch.index.mapper.IpFieldMapper$IpFieldType (IpFieldMapper.java:247)
```

File: server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java [hunk 2]
```diff
@@ -222,25 +224,42 @@ protected Object parseSourceValue(Object value) {
...
+                    query = InetAddressPoint.newPrefixQuery(name(), cidr.v1(), cidr.v2());
+                } else {
+                    InetAddress address = InetAddresses.forString(term);
+                    query = InetAddressPoint.newExactQuery(name(), address);
+                }
+            }
+            if (isSearchable() && hasDocValues()) {
+                String term = value.toString();
+                if (value instanceof String)
+                    value =  new BytesRef(((String) value).getBytes());
```
Comment (msfroh) [inline server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java:247]: `BytesRef` has a constructor that takes `CharSequence`. You can just say `value = new BytesRef((String)value)`.

File: server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java [hunk 3]
```diff
@@ -222,25 +224,42 @@ protected Object parseSourceValue(Object value) {
...
                     final Tuple<InetAddress, Integer> cidr = InetAddresses.parseCidr(term);
                     return InetAddressPoint.newPrefixQuery(name(), cidr.v1(), cidr.v2());
                 }
-                InetAddress address = InetAddresses.forString(term);
-                return InetAddressPoint.newExactQuery(name(), address);
+                return new IndexOrDocValuesQuery(query, SortedSetDocValuesField.newSlowExactQuery(name(), (BytesRef)(value)));
             }
+            if (hasDocValues()) {
+                return SortedSetDocValuesField.newSlowExactQuery(name(), ((BytesRef) value));
+            }
```
Comment (msfroh) [inline server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java:256]: If you get here, then you haven't passed through the logic above that converts a String value to `BytesRef`, so you will probably get a `ClassCastException`.

File: server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java (hunk 2 above)
Comment (harshavamsi) [reply server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java]: ah, okay

**Comment by harshavamsi** [comment]
Failures are flaky tests. @msfroh @reta another stab?

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/340_doc_values_field.yml [hunk 4]
```diff
@@ -1,8 +1,8 @@
 setup:
   - skip:
       features: [ "headers" ]
-      version: " - 2.11.99"
-      reason: "searching with only doc_values was added in 2.12.0"
+      version: " - 2.99.99"
+      reason: "searching with only doc_values was added in 3.0.0"
```
Comment (msfroh) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search/340_doc_values_field.yml:5]: I don't think I noticed this on the earlier commits, but this should only be scoped to the DV-only cases. As long as the fields are indexed, the queries will still work on earlier versions.

Comment (harshavamsi) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/340_doc_values_field.yml:5]: @msfroh we revert these on both main and 2.12 branches so do you think it still matters?

File: CHANGELOG.md [hunk 5]
```diff
@@ -209,6 +209,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Use slice_size == shard_size heuristic in terms aggs for concurrent segment search and properly calculate the doc_count_error ([#11732](https://github.com/opensearch-project/OpenSearch/pull/11732))
 - Added Support for dynamically adding SearchRequestOperationsListeners with SearchRequestOperationsCompositeListenerFactory ([#11526](https://github.com/opensearch-project/OpenSearch/pull/11526))
 - Ensure Jackson default maximums introduced in 2.16.0 do not conflict with OpenSearch settings ([#11890](https://github.com/opensearch-project/OpenSearch/pull/11890))
+- Extract cluster management for integration tests into JUnit test rule out of OpenSearchIntegTestCase ([#11877](https://github.com/opensearch-project/OpenSearch/pull/11877))
```
[the remark this answers is not in the document]
Comment (harshavamsi) [reply CHANGELOG.md]: Whoops, reverted that change.

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/340_doc_values_field.yml (hunk 4 above)
Comment (harshavamsi) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/340_doc_values_field.yml]: @msfroh updated this to only skip for the doc_values case.

**Review by msfroh — APPROVED** [review]

**Comment by reta** [comment]
@andrross LGTY? :-)

**Review by andrross — APPROVED** [review]

**Review by msfroh — APPROVED** [review]

**Comment by mkhludnev** [comment]
I'm afraid there is an edge case https://github.com/opensearch-project/OpenSearch/commit/cc9ee9d572dc395f6de81fd4d393fe9f1962fa50#r148997353

**Comment by msfroh** [comment]
> I'm afraid there is an edge case

@mkhludnev  -- are you able to fix that as part of https://github.com/opensearch-project/OpenSearch/pull/16391 ? If not, we can open a separate issue for it.

**Comment by mkhludnev** [comment]
#16628 here we go. #16391 turns out deeper than I though. tbc

**Pull request merged**